File: python/twoSum.py
# https://leetcode.com/problems/two-sum/

from typing import Dict, List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        Find two numbers that add up to a target value.

        Complexity:
        - Time: O(n)
        - Space: O(n)
        """
        # Keep track of the numbers we have seen so far and their indices
        seen_numbers: Dict[int, int] = {}

        # Return the index pairing that is expected for this function
        for idx, num in enumerate(nums):
            # Calculate the complement that we need to find
            complement = target - num

            # If the complement is already in the dictionary, we found our pair
            if complement in seen_numbers:
                return [seen_numbers[complement], idx]

            # Otherwise, store the current number with its index
            seen_numbers[num] = idx

        raise RuntimeError("There should be exactly one solution")

File: java/TwoSum.java
// https://leetcode.com/problems/two-sum/

import java.util.HashMap;

public class TwoSum {
    /**
     * Time: O(n)
     * Space: O(n)
     *
     * We create a mapping of the numbers we have seen so far. While
     * we do this, we also check if the complement of the current
     * number (i.e., target - nums[i]) exists in the map. If it
     * does, we return the indices of the two numbers that add up to
     * the target.
     */
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> seenValues = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (seenValues.containsKey(complement)) {
                return new int[]{seenValues.get(complement), i};
            }
            seenValues.put(nums[i], i);
        }
        throw new IllegalArgumentException("This should not happen");
    }
}
